pycdlib-explorer: Add functionality for modifying files in place

ticky opened the pull request:
This exposes the ability to modify a file in-place in `pycdlib-explorer`.

To facilitate this:
- `do_ls`: now shows file lengths in bytes alongside file names
- `do_modify_file_in_place`: calls the `modify_in_place` API to replace a file, uses the file's on-disk length
- `main`: accept an additional "--read-write" parameter to open the target file in `r+b` mode, so that `modify_file_in_place` can work

Note that this currently includes the content of #70 as I needed it for some of my testing. I am happy to rebase this if and when that's merged.

clalancette commented:
> Note that this currently includes the content of #70 as I needed it for some of my testing. I am happy to rebase this if and when that's merged.

I've now merged #70, so please go ahead and rebase it when you have a chance.  Thanks!

ticky commented:
I've rebased this, and after having had some success with this initially I've started seeing inconsistent behaviour from this with an ISO9660+UDF disk image, wherein replacing a file creates an ISO that `pycdlib` itself then fails to open:

```
Traceback (most recent call last):
  File "/opt/homebrew/bin/pycdlib-explorer", line 620, in <module>
    main()
  File "/opt/homebrew/bin/pycdlib-explorer", line 604, in main
    iso.open_fp(fp)
  File "/opt/homebrew/lib/python3.9/site-packages/pycdlib/pycdlib.py", line 4208, in open_fp
    self._open_fp(fp)
  File "/opt/homebrew/lib/python3.9/site-packages/pycdlib/pycdlib.py", line 2360, in _open_fp
    le_ptrs, extent_to_ptr = self._parse_path_table(self.pvd.path_table_size(),
  File "/opt/homebrew/lib/python3.9/site-packages/pycdlib/pycdlib.py", line 1257, in _parse_path_table
    ptr.parse(data[offset:offset + read_len])
  File "/opt/homebrew/lib/python3.9/site-packages/pycdlib/path_table_record.py", line 60, in parse
    self.parent_directory_num) = struct.unpack_from(self.FMT, data[:8], 0)
struct.error: unpack_from requires a buffer of at least 8 bytes for unpacking 8 bytes at offset 0 (actual buffer size is 4)
```

I've seen other errors from a file modified in what appears to be the same way:

```
Traceback (most recent call last):
  File "/opt/homebrew/bin/pycdlib-explorer", line 620, in <module>
    main()
  File "/opt/homebrew/bin/pycdlib-explorer", line 604, in main
    iso.open_fp(fp)
  File "/opt/homebrew/lib/python3.9/site-packages/pycdlib/pycdlib.py", line 4208, in open_fp
    self._open_fp(fp)
  File "/opt/homebrew/lib/python3.9/site-packages/pycdlib/pycdlib.py", line 2380, in _open_fp
    ic_level, lastbyte = self._walk_directories(self.pvd, extent_to_ptr,
  File "/opt/homebrew/lib/python3.9/site-packages/pycdlib/pycdlib.py", line 1092, in _walk_directories
    rr = new_record.parse(vd, data[offset:offset + lenbyte],
  File "/opt/homebrew/lib/python3.9/site-packages/pycdlib/dr.py", line 226, in parse
    raise pycdlibexception.PyCdlibInvalidISO('Little-endian (%d) and big-endian (%d) extent location disagree' % (extent_location_le, utils.swab_32bit(extent_location_be)))
pycdlib.pycdlibexception.PyCdlibInvalidISO: Little-endian (1060) and big-endian (14848) extent location disagree
```

I then re-checked and the one I'd created when I initially got it working has similar problems with not being able to be opened:

```
Traceback (most recent call last):
  File "/opt/homebrew/bin/pycdlib-explorer", line 620, in <module>
    main()
  File "/opt/homebrew/bin/pycdlib-explorer", line 604, in main
    iso.open_fp(fp)
  File "/opt/homebrew/lib/python3.9/site-packages/pycdlib/pycdlib.py", line 4208, in open_fp
    self._open_fp(fp)
  File "/opt/homebrew/lib/python3.9/site-packages/pycdlib/pycdlib.py", line 2380, in _open_fp
    ic_level, lastbyte = self._walk_directories(self.pvd, extent_to_ptr,
  File "/opt/homebrew/lib/python3.9/site-packages/pycdlib/pycdlib.py", line 1092, in _walk_directories
    rr = new_record.parse(vd, data[offset:offset + lenbyte],
  File "/opt/homebrew/lib/python3.9/site-packages/pycdlib/dr.py", line 237, in parse
    raise pycdlibexception.PyCdlibInvalidISO('Little-endian and big-endian seqnum disagree')
pycdlib.pycdlibexception.PyCdlibInvalidISO: Little-endian and big-endian seqnum disagree
```

In the latter case I replaced one file (the others I replaced two) with an identically sized modified version, and this version mounts successfully under macOS. The others appear as volumes with no files when mounted.

This was running a copy of the previous stable version with my modifications applied, though I can't see any obvious changes since which would affect this. I'm not sure if this means my code for this new command is bad, or if there's an underlying bug?

This disk image is not one I can distribute publicly, is there anything else I can do to assist with debugging this?

Hope you can help! :)

clalancette commented:
There is some possibility that this is a pre-existing bug; the UDF support in pycdlib is still pretty basic.  One thing that would help debug is to generate an ISO without UDF, and see if that is any better.  That will at least tell us if the problem is in `modify_in_place`, or in the UDF support.

clalancette reviewed (approved):
I went through and simplified this PR a bit.  In particular, I got rid of the command-line argument in favor of just always opening the ISO in r+b mode.  I also got rid of all of the extra arguments in the `modify_file_in_place` command; they are all actually ignored at this time anyway.  Finally, I used a context manager for opening the source file.  With all of that in place, I'm happy with this now, so I'm going to go ahead and merge it.  I'm not sure what is happening with your other errors, so if you'd like to follow up with that please open a separate issue.

Thanks for the contribution!
